lib/python/picongpu/plugins/data/base_reader.py

```diff
@@ -66,7 +66,7 @@ def get_iterations(self, **kwargs):
         """
         raise NotImplementedError
 
-    def get(self, **kwargs):
+    def get(self, *args, **kwargs):
         """
         Parameters
```

Comment:
We should document `species` and `species_filter` here as well or point to the according method that this forwards to. Otherwise users cannot find their signature when checking `.get?`

Reply:
This might be tricky since this function passes the `*args` and `**kwargs` on to the `_get_for_iteration` method of the derived classes which do not necessarily have common arguments that could be mentioned here. 
Maybe python has some tricks that allow displaying the docstring of `eh_data._get_for_iteration` when calling `help(eh_data.get)` instead of the docstring for `BaseReader.get` but I am not familiar with any. Do you happen to know?

Reply:
@ax3l Are you fine with that?

Reply:
@codingS3b I think you should be able to "monkeypatch" (actual term) the `.__doc__` method in your derived class for this purpose.

Reply:
After playing around a little bit, that is indeed possible. I tried it using the following example:
```py
def modify_doc(doc):
    def f_with_doc(f):
        f.__doc__ = doc
        return f
    return f_with_doc


class Base(object):
    def _get(self, *args, **kwargs):
        """
        Actual implementation, not given in base class
        """
        raise NotImplementedError
    
    def get(self, *args, **kwargs):
        """
        Not useful docstring, refer to derived _get
        """
        # NOTE: when calling help on that function, we should display help for _get maybe?
        print("Base: common stuff all derived classes need")
        
        #self.get.__doc__ == self._get.__doc__
        #self.get.__func__.__doc__ = self._get.__doc__
        return self._get(*args, **kwargs)
    
class Derived(Base):
    def __init__(self):
        super().__init__()
    def _get(self, a, b=2):
        """
        Documentation of the derived class _get
        Parameters
        ----------
        a: int
            an integer
        b: int
            a second integer
        """
        
        print("Derived: a=", a, "b=", b)
    
    @modify_doc(Derived._get.__doc__)
    def get(self, *args, **kwargs):
        return super().get(*args, **kwargs)

a = Base()
help(a.get)
# Help on method get in module __main__:

# get(*args, **kwargs) method of __main__.Base instance
#    Not useful docstring, refer to derived _get

d = Derived()
help(d.get)
#Help on method get in module __main__:
#
#get(*args, **kwargs) method of __main__.Derived instance
#    Documentation of the derived class _get
#    Parameters
#    ----------
#    a: int
#        an integer
#    b: int
#        a second integer
```
However, this has the downside that we have to add the decorated `get` function to each derived class, only for the purpose of documentation, without adding functionality and simply calling the base class function. Maybe just adding the remark to have a look at the derived classes function signature is simpler...

Reply:
Cool I think you can avoid implementing a `.get` in each derived class by [monkey-patching](https://filippo.io/instance-monkey-patching-in-python/).
Not sure if that makes your code shorter as it already is, though.

Reply:
Thanks for sharing the link. I'm not sure I get you right.
Staying with the above example, do you mean something like 
```python
d.get.__doc__ = "new documentation"
# throws AttributeError: attribute '__doc__' of 'method' objects is not writable

# however this seems to work
d.__class__.get.__doc__ = "new documentation
```
? When should this monkey patching take place then? During object construction or already at the time of class definition?

Reply:
About the when I have no good enough experience with, preferably the earlier the better. Anyway, at object construction time sounds still good enough since a user will ask the member function of an already instantiated object for its doc string.

Reply:
I added an `_update_docstrings` function in the constructor of the base reader class that replaces the docstring of the `get` function with the one from `_get_for_iteration` function of the derived class. The minor problem is now that this replacement hides the option to pass a `time=` argument to the `get` function since the `_get_for_iteration` only accepts `iteration` as argument and the base class method takes care of converting time to iteration.

Reply:
@codingS3b Did you solve this issue?